File: src/NovelRT.Java/src/main/java/com/github/novelrt/event/Event.java
package com.github.novelrt.event;

import com.github.novelrt.internal.HandleDeleter;
import com.github.novelrt.internal.HandleObject;
import com.github.novelrt.internal.UsedNatively;

import java.util.*;

public abstract class Event<T extends EventListener> extends HandleObject {
  private final Set<T> listeners = new LinkedHashSet<>();

  protected Event(long handle) {
    super(handle, false, HandleDeleter.NONE);
  }

  // TODO: Use Guava's ImmutableSet<T> because UGH unmodifiableSet...
  public Set<T> getListeners() {
    return Collections.unmodifiableSet(listeners);
  }

  public void subscribe(T listener) {
    if (listeners.contains(listener)) {
      return;
    }

    addSubscription(listener);
    listeners.add(listener);
  }

  public void unsubscribe(T listener) {
    if (!listeners.contains(listener)) {
      return;
    }

    removeSubscription(listener);
    listeners.remove(listener);
  }

  @UsedNatively
  private void onRemoval(T listener) {
    listeners.remove(listener);
  }

  protected abstract void addSubscription(T listener);
  protected abstract void removeSubscription(T listener);
}

File: src/NovelRT.Java/core/src/main/kotlin/com/github/novelrt/event/Event.kt
package com.github.novelrt.event

abstract class Event<T> internal constructor() {
    protected val eventListeners: MutableSet<T> = LinkedHashSet()

    operator fun plusAssign(listener: T) {
        subscribe(listener)
    }

    operator fun minusAssign(listener: T) {
        unsubscribe(listener)
    }

    private fun subscribe(listener: T): T {
        if (listener in eventListeners) {
            return listener
        }
        eventListeners.add(listener)
        return listener
    }

    private fun unsubscribe(listener: T) {
        if (listener !in eventListeners) {
            return
        }
        eventListeners.remove(listener)
    }

    protected inline fun fireEvent(runner: (listener: T) -> Unit) {
        for (listener in eventListeners) {
            runner(listener)
        }
    }
}
